File: day8-1/go/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
)

func main() {
	if err := run(); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}

func run() error {
	input, err := readInput("input.txt")
	if err != nil {
		return err
	}

	sum, err := accBeforeLoop(input)
	if err != nil {
		return err
	}
	fmt.Printf("%d\n", sum)

	return nil
}

func readInput(path string) (string, error) {
	data, err := ioutil.ReadFile(path)
	return string(data), err
}

func accBeforeLoop(input string) (int, error) {
	v := newVm(input)
	pcs := map[int]bool{}
	for {
		if err := v.Next(); err != nil {
			return 0, err
		}
		pc := v.PC()
		if pcs[pc] {
			return v.Acc(), nil
		}
		pcs[pc] = true
	}
}

func newVm(input string) *vm {
	return &vm{
		ops: strings.Split(strings.TrimSpace(input), "\n"),
	}
}

type vm struct {
	ops []string
	pc  int
	acc int
}

func (v *vm) Next() error {
	op := v.ops[v.pc]
	parts := strings.Split(op, " ")
	if len(parts) != 2 {
		return fmt.Errorf("bad op: %s", op)
	}
	n, err := strconv.ParseInt(parts[1], 10, 64)
	if err != nil {
		return fmt.Errorf("bad op: %s", op)
	}
	switch parts[0] {
	case "nop":
		v.pc++
	case "acc":
		v.acc += int(n)
		v.pc++
	case "jmp":
		v.pc += int(n)
	}
	return nil
}

func (v *vm) PC() int {
	return v.pc
}

func (v *vm) Acc() int {
	return v.acc
}

File: day8-1/rust/src/main.rs
use std::collections::HashMap;

fn main() {
    let acc = get_acc("
nop +0
acc +1
jmp +4
acc +3
jmp -3
acc -99
acc +1
jmp -4
acc +6
");

    println!("{}", acc);
}

fn get_acc(input: &str) -> isize {
    let ops: Vec<&str> = input.trim().split("\n").collect();
    let mut prev_pcs = HashMap::new();

    let mut pc: isize = 0;
    let mut acc: isize = 0;
    loop {
        let op = ops[pc as usize];
        let parts: Vec<&str> = op.split(" ").collect();
        if let [op, n_str] = parts[..] {
            let n = n_str.parse::<isize>().expect(n_str);
            match op {
                "nop" => pc += 1,
                "acc" => {
                    acc += n;
                    pc += 1
                },
                "jmp" => {
                    pc += n
                }
                _ => println!("unknown op: {}", op)
            }
            if prev_pcs.contains_key(&pc) {
                return acc
            }
            prev_pcs.insert(pc, true);
        } else {
            println!("bad op: {}", op)
        }
    }
}
